rr_graph_uxsdcxx.h

```diff
@@ -0,0 +1,398 @@
+ *
+ * Cmdline: ./uxsdcxx.py rr_graph.xsd
+ * Input file: rr_graph.xsd
+ * md5sum of input file: 2b7bc77a1e7c006250cf0351f8046f9e
+ */
+
+/* All uxsdcxx functions and structs live in this namespace. */
+namespace uxsd {
+
+typedef const uint32_t __attribute__((aligned(1))) triehash_uu32;
+typedef const uint64_t __attribute__((aligned(1))) triehash_uu64;
+static_assert(alignof(triehash_uu32) == 1, "Unaligned 32-bit access not found.");
+static_assert(alignof(triehash_uu64) == 1, "Unaligned 64-bit access not found.");
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#define onechar(c, s, l) (((uint64_t)(c)) << (s))
+#else
+#define onechar(c, s, l) (((uint64_t)(c)) << (l-8-s))
+#endif
+
+
+/* Stores a vector of elements in a shared pool. Consists
+ * of an offset into the pool and a size. It's faster than a regular
+ * vector, but one can only insert into it when it's the last vector
+ * in the pool. */
+template<class T, std::vector<T> &pool>
+class collapsed_vec {
+private:
+	uint32_t _size;
+	uint32_t _offset;
+public:
+	inline collapsed_vec(){
+		_size = 0;
+		_offset = pool.size();
+	}
+	inline T& back(){
+		return pool[_offset+_size-1];
+	}
+	inline T* begin(){
+		return &pool[_offset];
+	}
+	inline T* end(){
+		return &pool[_offset+_size];
+	}
+	inline T& operator[](uint32_t i){
+		return pool[_offset+i];
+	}
+	inline void push_back(const T &x){
+		assert(_size+_offset == pool.size());
+		pool.push_back(x);
+		_size++;
+	}
+	inline uint32_t size(){
+		return _size;
+	}
+};
+
+/* A pool of strings. It exposes an add(const char *x) function
+ * which is like a better strdup.
+ * add() calls strlen on the argument to see how much space is needed.
+ * Then it checks its vector of chunks to see if there is a chunk it
+ * would fit into. If none is found, a new one is allocated with a size of
+ * max(length of string, last chunk's size*2). */
+class string_pool_impl {
+private:
+	const uint32_t INITIAL_SIZE = 1024;
+	struct chunk {
+		uint32_t used;
+		uint32_t size;
+		char * mem;
+	};
+	std::vector<chunk> chunks;
+public:
+	string_pool_impl(){
+		chunk c;
+		c.used = 0;
+		c.size = INITIAL_SIZE;
+		c.mem = (char *)std::malloc(INITIAL_SIZE);
+		chunks.emplace_back(c);
+	}
+	~string_pool_impl(){
+		for(auto &c: chunks)
+			free(c.mem);
+	}
+	const char *add(const char *x){
+		uint32_t len = std::strlen(x)+1;
+		char *out;
+		for(auto &c: chunks){
+			if(c.used+len <= c.size){
+				out = &c.mem[c.used];
+				c.used += len;
+				std::memcpy(out, x, len);
+				return out;
+			}
+		}
+		chunk n;
+		n.used = 0;
+		n.size = std::max(chunks.back().size*2, len);
+		n.mem = (char *)std::malloc(n.size);
+		std::memcpy(n.mem, x, len);
+		chunks.emplace_back(n);
+		return n.mem;
+	}
+	void clear(){
+		for(int i=1; i<chunks.size(); i++){
+			free(chunks[i].mem);
+		}
+		chunks.resize(1);
+		chunks[0].used = 0;
+		std::memset(chunks[0].mem, 0, INITIAL_SIZE);
+	}
+};
+/* Forward decl of generated data types. Needed for the pools. */
+struct t_channel;
```

Comment:
How are these declarations sorted?

Reply:
According to "tree height", from schema file appearance order: https://github.com/duck2/uxsdcxx/blob/master/uxsdcxx.py#L270 This is done to put the "root element" last, so that we can embed other structs in it.

Should I generate a comment about this?

Reply:
Yes.